DRY conditions.go / small perf boost

andrewvc opened the pull request:
conditions.go was checking the configured conditions with conditional branching both at
config load time and once per event. This patch removes the redundant per-event checks.

There may be a small perf boost with this patch as well.

urso commented:
I'm pretty sure this change will boost performance a lot.

A more idiomatic solution would use interfaces instead of a function pointer. E.g. change `Condition` type to `type Condition interface` and create one implementation per condition type.

Benchmarks would be nice to have for comparing future changes as well.

andrewvc commented:
@ruflin @urso thanks for the feedback, glad to make both refactors. I like your proposal of one concrete implementation per condition type. The current logic, based on which fields happen to be set, is harder to understand.

andrewvc commented:
This is an in-progress commit, I still have to run benchmarks against it and maybe add more tests. A few small checks may also be failing. It would be great to have feedback on the general direction here at this point however.

This PR is now *slightly* larger , but is, IMHO, much better, having taken @urso 's suggestion of using an interface rather than a function pointer. 

There was a lot of logic in the `conditionals.go` file which is now better broken up, more readable, and function pointer free. Porting these functions to interfaces and leaving them in a monolithic file felt like a waste of effort, since it'd still be hard to read. 

This is a pretty big refactor, I tried to limit it to mostly moving code around around, and kept actual refactors to a bare minimum, so while the number of lines is large, the amount of new logic is minimal. I did have to add a lot of comments to old functions to make Hound happy.

A key change here is the creation of a new `conditionals` package, which encapsulates the conditionals logic. Keeping it all in `processors` seemed like not a great idea. There's more than enough code in conditionals to justify a new package IMHO.

Another significant change was moving the conditional objects to be values rather than pointer receivers to following the style recommendations on the [golang code review comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type). `Conditional` concrete types are all either immutable maps , tiny structs, or slices. I'm a golang newbie, so maybe I'm misunderstanding that advice, but it seems like anything large is hidden away in a map or a slice, and the remaining copied values are small.

Looking over the code, we don't have great test coverage here at the unit level, it might be nice to add some more tests, but I'm fine leaving the test coverage as is, which is to say, largely at the functional level. Thoughts @urso && @ruflin ? I think it would be fair to say that's out of scope for this PR at the very least, this is already a quite large one.

ruflin commented on libbeat/processors/conditions/and.go:
```diff
@@ -0,0 +1,44 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package conditions
```
+1 on having a separate package and each in one file. I was even thinking if we should have each condition in it's own package but that would probably be overkill. Somehting like `processors/condition/and/*`

andrewvc replied:
++ on keeping it small

ruflin commented on libbeat/processors/conditions/condition_fields.go:
```diff
@@ -0,0 +1,58 @@
+import "fmt"
+
+// ConditionFields represents an arbitrary map in a config file.
+type ConditionFields struct {
```
I would call it just `Fields` to not repeat the package name.

andrewvc replied:
Yes, makes sense, will do

ruflin commented on libbeat/processors/conditions/equals.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// EqualsCondition is a Condition for testing string equality.
+type EqualsCondition map[string]equalsValue
```
Would also call it `Equals` here and leave out the `condition` part as it's part of the package name. If someone instantiates it from outside it reads `conditions.NewEquals`

andrewvc replied:
++

ruflin commented on libbeat/processors/conditions/conditions.go:
```diff
@@ -0,0 +1,109 @@
+
+// Condition is the interface for all defined conditions
+type Condition interface {
+	Execute(event ValuesMap) bool
```
No strong preference on my end but I'm curious why you renamed it from Check to Execute?

andrewvc replied:
It's not a super-strong preference on my side either, but I have a slight preference for execute. 

Check to me sounds like something that checks internal state. Execute sounds better IMHO in the context of an interpreter for an IR/AST.

That said, neither one is right, glad to switch back to Check if you or @urso have a strong preference there, it's a very minor thing.

urso replied:
I'd rather use check or something more descriptive. In Go anything implementing `Execute` can be a conditional, as the type checker checks for structural compatibility of values and interface. To me the name `Execute` implies an implementation can do "much more".

andrewvc replied:
Cool, I'll switch back to `Check`

andrewvc commented on libbeat/processors/conditionals.go:
```diff
@@ -0,0 +1,110 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package processors
```
This package is a bit of an awkward grab-bag, but I think it's still an improvement over the current code, and I don't want to take this refactor too far.

ruflin replied:
Agree, we can still do follow up PR's.

andrewvc commented:
OK, added some benchmarks, tested between master / this branch. It's a nice little boost for this part of the code, but at the end of the day, I doubt users will perceive a perf improvement here unless it turns out that the efficiency here improves the global story for CPU caching, which is possible, but probably not worth spending time to benchmark (I also doubt it will amount to anything anyone will care about).

With the benchmark data we do have, in the best case, you save 10ns per condition, which is about 10ms per million events with one condition. If you have a lot of those conditions that *could* add up. However, that savings is relatively smaller as the statements grow in complexity.

Raw benchmark data below. This is comparing the benchmarks in the old branch (which I added in https://github.com/andrewvc/beats/blob/dry-cond-old-bench/libbeat/processors/condition_test.go) with the new branch in ( https://github.com/andrewvc/beats/blob/dry-cond/libbeat/processors/conditions/conditions_test.go ).

The simple condition test tests a `has_fields` condition, which is about as simple as it gets. The `CombinedCondition` test tests more complexity with a compound statement.

```
benchmark                        old ns/op     new ns/op     delta
BenchmarkSimpleCondition-8       34.9          21.5          -38.40%
BenchmarkCombinedCondition-8     149           127           -14.77%
```

We should still do this, since the code is much better organized and easy to work with in this PR, and the perf is clearly better.

ruflin commented:
For me the biggest win of this change is the actual refactoring with the interfaces as it should make adding future conditions easier.

urso commented:
Nice code cleanups. I love finally having an extra package. Conditions are not only used by processors. You might consider to move the package one level up.

The tests still need some cleanup. Current test code makes it quite hard to localize and debug an actual failing test.  E.g. have tests in file per condtion (e.g. `and.go` with tests in `and_test.go`). Plus use `t.Run` for sub tests. So it's easier to find and run the exact test that failed.

Common pattern for table driven tests:

```
func TestX(t *testing.T) {
  type testCase struct {
    // test input, expected output, ...
  }

  cases := map[string]testCase{ // using map here ensures tests are run in random order
    "must succeed": testCase{ ... },
    "must fail": testCase{...}, 
  }

  for name, test := range cases {
    // common global setup/teardown

    test := test // capture loop variable in case tests are run in parallel (via t.Parallel())
    t.Run(name, func(t *testing.T) {
      // test driver code
    })
  }
}
```

Using this pattern, the test name will appear in logs + one can run tests selectively via `go test -run 'X/fail/'`. The `-run` arguments accepts a regular expressions with `/` delimiting test levels.

urso commented:
I'm not too worried about the benchmarks here (users can end up with tons of conditionals). The refactoring makes the code much more accessible and maintainable.

urso commented on libbeat/processors/conditions/conditions.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	if err != nil {
+		return nil, err
+	}
```
This kind of switch acts like a big `if-else-if-else` block.

I'm not a big fan of the Config object as is. At least we should have some validation no 2 fields in Config are set at the same time.

andrewvc replied:
I agree, however I think it makes sense to keep the scope of this PR down to moving code around as much as possible. I think we should change the behavior here in a follow-up PR that I'm glad to write.

andrewvc replied:
To be clear, this is existing code, not new, I just moved it from the old location with the minimal number of changes.

andrewvc commented:
@urso great point WRT to the tests, I'll split them out later today. Thanks for the advice on test structure as well.

andrewvc commented:
I've made the following changes:

1. Refactored the various tests out to separate files + broke up the big combination tests into individual ones. Ultimately, it made more sense to do a function per-test than to go the anonymous route.
2. Renamed conditions.Execute to conditions.Check

I think that we should do some of the additional work in follow up PRs. This means:

1. Refactoring `conditions.NewCondition` to remove the big switch and validate in a more robust way.
2. Potentially making the tests more comprehensive. As I've split out the tests I've noticed that some conditions are tested better than others. I'd like to hear what @urso and @ruflin think about the current state of the tests, but they are at the moment spot checks, and not comprehensive. For this core functionality I think it makes sense to test a range of different conditions instead of the current spot checks.

I would prefer to do this in separate PRs because this one is unwieldy and I'd like to constrain its scope to reorganization. While there is some refactoring in here, it is in order to facilitate reorganization, not to change any actual logic besides the core goal of removing the once-per-event branching on condition configuration.

andrewvc commented:
I also just moved the package up a level to be a sibling of processors as @urso suggested.

ruflin commented:
@andrewvc There seems to be something off with the formatting. `make fmt` should solve it.

andrewvc commented on libbeat/processors/conditionals_test.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	for i, test := range tests {
+		t.Logf("run test (%v): %v", i, test.title)
```
As discussed, this can be deferred to a future PR.

andrewvc commented:
Squashed + all green. Ready to merge when y'all are :)

ruflin commented:
@andrewvc Thanks for the improvements and cleanup. 🎉
